### PR TITLE
feat(generic-metrics): Add unique constraint to indexer

### PR DESCRIPTION
**Pull request by ayirr7**
The follow-up step to https://github.com/getsentry/sentry/pull/46404. Here, we're modifying the uniqueness constraint to include the newly-added `use_case_id` column.

### CODE REVIEW

**Review by wedamija — APPROVED** [review]

**Review by wedamija — CHANGES_REQUESTED** [review]
Actually... I just realised that I think this will lock the table since it's not going to add the unique constraint concurrently.

You can do this concurrently via
```
CREATE UNIQUE INDEX CONCURRENTLY <constraint_name> ON <table_name> USING btree (<column_one>, <column_two>);
```

Then 
```
ALTER TABLE <table_name> ADD CONSTRAINT <constraint_name> UNIQUE USING INDEX <name_of_index_created_above>
```

More detail here: https://adriennedomingus.medium.com/adding-a-unique-together-constraint-concurrently-b364b9f3eb38

**Comment by wedamija** [comment]
> Actually... I just realised that I think this will lock the table since it's not going to add the unique constraint concurrently.
> 
> You can do this concurrently via
> 
> ```
> CREATE UNIQUE INDEX CONCURRENTLY <constraint_name> ON <table_name> USING btree (<column_one>, <column_two>);
> ```
> 
> Then
> 
> ```
> ALTER TABLE <table_name> ADD CONSTRAINT <constraint_name> UNIQUE USING INDEX <name_of_index_created_above>
> ```
> 
> More detail here: https://adriennedomingus.medium.com/adding-a-unique-together-constraint-concurrently-b364b9f3eb38

(I'm not totally sure, but if you skip the custom sql and just use the Django operations, our framework might do this safely for you. You could try and see what sql is generated)

File: src/sentry/migrations/0397_add_unique_constraint_indexer.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                        fields=("string", "organization_id", "use_case_id"),
+                        name="perf_unique_org_string_usecase",
+                    ),
+                ),
+                migrations.RemoveConstraint(
+                    model_name="perfstringindexer",
+                    name="perf_unique_org_string",
+                ),
+            ],
+        )
```
Comment (wedamija) [inline src/sentry/migrations/0397_add_unique_constraint_indexer.py:41]: You'll need to move these operations outside of `SeparateDatabaseAndState`. Having them in `state_operations` causes them to only update the app state, and not change the database.

Comment (ayirr7) [reply src/sentry/migrations/0397_add_unique_constraint_indexer.py]: Thanks for catching that

Comment (ayirr7) [reply src/sentry/migrations/0397_add_unique_constraint_indexer.py]: the generated SQL looks good, though it doesn't have the USING btree clause

Comment (wedamija) [reply src/sentry/migrations/0397_add_unique_constraint_indexer.py]: I think this is fine. I think the validation pattern @markstory mentioned above only really applies to other types of constraints, like `NOT NULL` etc. Adding the unique index means that inherently there are no duplicates on the table, so I think that all that the constraint needs to do is verify that the unique is in place.

**Review by wedamija — APPROVED** [review]

**Pull request merged**